fluent/fluent.go

```diff
@@ -323,17 +370,32 @@ func (f *Fluent) write(data []byte) error {
 		}
 		f.muconn.Unlock()
 
-		// We're connected, write data
+		// We're connected, write msg
 		t := f.Config.WriteTimeout
 		if time.Duration(0) < t {
 			f.conn.SetWriteDeadline(time.Now().Add(t))
 		} else {
 			f.conn.SetWriteDeadline(time.Time{})
 		}
-		_, err := f.conn.Write(data)
+		_, err := f.conn.Write(msg.data)
 		if err != nil {
 			f.close()
 		} else {
+			// Acknowledgment check
+			if msg.ack != "" {
+				resp := &AckResp{}
+				if f.Config.MarshalAsJSON {
+					dec := json.NewDecoder(f.conn)
+					err = dec.Decode(resp)
+				} else {
+					r := msgp.NewReader(f.conn)
+					err = resp.DecodeMsg(r)
+				}
+				if err != nil || resp.Ack != msg.ack {
```

Comment:
It seems to have to close `f` always, not only when ack error happens. Is it correct?
If so, we should call `f.close()` before this `if` clause.

Reply:
I'm not sure about this one. It seems to me that in the original code `f` was closed only on errors. For normal operation the connection was reused.

Or did I miss something?

Reply:
Ah, your comment is true! My comment is based on wrong understanding.

Reply:
But anyway, `err` is `nil` always here because this line is in `else` for `if err != nil`.

Reply:
(I made one more incorrect comment just before, and deleted it)